File: algorithms/Radix Sort/Python/code.py
def counting_sort(arr, exponent):
    size = len(arr)
    output = [0] * size
    count = [0] * 10

    # Count occurrences of each digit.
    for value in arr:
        digit = (value // exponent) % 10
        count[digit] += 1

    # Convert counts into positions.
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Build the output array from right to left.
    # This preserves stability.
    for i in range(size - 1, -1, -1):
        digit = (arr[i] // exponent) % 10

        output[count[digit] - 1] = arr[i]
        count[digit] -= 1

    # Copy the sorted values back.
    for i in range(size):
        arr[i] = output[i]


def radix_sort(arr):
    if len(arr) <= 1:
        return

    max_value = arr[0]

    # Find the largest value.
    for value in arr:
        if value > max_value:
            max_value = value

    # Process each digit from least significant
    # to most significant.
    exponent = 1

    while max_value // exponent > 0:
        counting_sort(arr, exponent)
        exponent *= 10


arr = [170, 45, 75, 90, 802, 24, 2, 66]

radix_sort(arr)

print(arr)

File: algorithms/Radix Sort/C/code.c
#include <stdio.h>
#include <stdlib.h>

void counting_sort(int arr[], int size, int exponent)
{
    int *output = malloc(size * sizeof(int));
    int count[10] = {0};

    if (output == NULL)
    {
        return;
    }

    /* Count occurrences of each digit. */
    for (int i = 0; i < size; i++)
    {
        int digit = (arr[i] / exponent) % 10;
        count[digit]++;
    }

    /* Convert counts into positions. */
    for (int i = 1; i < 10; i++)
    {
        count[i] += count[i - 1];
    }

    /*
     * Build the output array from right to left.
     * This preserves stability.
     */
    for (int i = size - 1; i >= 0; i--)
    {
        int digit = (arr[i] / exponent) % 10;

        output[count[digit] - 1] = arr[i];
        count[digit]--;
    }

    /* Copy the sorted values back into the input array. */
    for (int i = 0; i < size; i++)
    {
        arr[i] = output[i];
    }

    free(output);
}

void radix_sort(int arr[], int size)
{
    if (size <= 1)
    {
        return;
    }

    int max = arr[0];

    /* Find the largest value. */
    for (int i = 1; i < size; i++)
    {
        if (arr[i] > max)
        {
            max = arr[i];
        }
    }

    /*
     * Process each digit from least significant
     * to most significant.
     */
    for (int exponent = 1; max / exponent > 0; exponent *= 10)
    {
        counting_sort(arr, size, exponent);
    }
}

int main(void)
{
    int arr[] = {170, 45, 75, 90, 802, 24, 2, 66};
    int size = sizeof(arr) / sizeof(arr[0]);

    radix_sort(arr, size);

    for (int i = 0; i < size; i++)
    {
        printf("%d ", arr[i]);
    }

    printf("\n");

    return 0;
}